### ISSUE TITLE
parsing model dump to get trees

### ISSUE THREAD

**Issue by AlexandraBomane** [dmlc/xgboost#3802]
I use the Python API (sklearn wrapper) to generate an XGBClassifier with n_estimators set to 1000. I'm trying to collect the trees built during the model fitting to see the features involved in the predictions.
I used model.get_booster().dump_model(with_stats=True) to dump the model into a text. Among the 1000 boosters, only 72 are decision trees that show splits involving features. The remaining boosters are single leaves (see #3205 ) indicating a value (the weight of the leaf, I guess).
I would like to know if I just need to consider the 72 decision trees as actual weak-learners involved in the predictions of sample labels, and then ignore the remaining boosters that correspond to single leaves. And I would like to know what do represent these single leaves regarding to the model.

Thanks in advance

**Comment by hcho3** [comment]
Did you perform early stopping? With early stopping, training would probably stop at 72 iteration mark.

**Comment by AlexandraBomane** [comment]
No, I directly fit the XGBClassifier without performing any early stopping.

**Comment by hcho3** [comment]
Can you plot the evaluation metric on training and validation data?

**Comment by AlexandraBomane** [comment]
I cannot generate such plot because I don't use training and validation data, I fit my model on my all data.

**Comment by hcho3** [comment]
> you said that single leaf boosters are ignored for the prediction

I did not say that. Single leaf trees will be used for prediction, although their impact on prediction are probably minimal. Check the magnitude of leaf outputs for single leaf trees and see if they are small.

Also, can you plot the evaluation metric on the data you fit your model on? Did the metric plateaued?

**Comment by AlexandraBomane** [comment]
How can I check the magnitude of leaf outputs for single leaf trees ?
What is the function to plot the evaluation metric ? I tried "evals_result()", but it doesn't work probably because I'm using the Sklearn wrapper and not xgboost.XGBClassifier.

**Comment by AlexandraBomane** [comment]
Do the single leaf trees are involved  in the prediction participating to the iterative corrections of trees leading to the prediction ?

**Comment by AlexandraBomane** [comment]
> Look at the return value of get_dump().
Taking the output of get_dump(), then should I sum all the absolute values leaves weights ? If it is the case, should I consider that if the result < 1, so the single-leaf trees do not have important influence on the prediction ?

> Did you pass eval_set to the fit() function?
No, I didn't pass `eval_set` to the `fit()` function.

> The question is the degree at which they influence the prediction?
Exactly.

**Comment by AlexandraBomane** [comment]
By the way, how it is possible to get single-leaf trees ?

**Comment by hcho3** [comment]
There are several parameter that control pruning process, e.g. `min_child_weight` and `delta`. You should change these.

**Comment by AlexandraBomane** [comment]
Thanks for your help,
Just some more questions : finally the single-leaf trees are due to pruning ? (as said also [here](https://stackoverflow.com/questions/35136404/xgboost-binary-logistic-regression)).
I'm used to see `n_estimator` set to 1000 in tutorials for XGBoost and Kaggle recommendations. Do you confirm that it is a good generic setting if I don't want to spend time on tuning ?

**Comment by hcho3** [comment]
Yes, pruning is a likely cause of single-leaf trees. You should combine parameter tuning (e.g. `min_child_weight` and `delta`) with early stopping to obtain a model that's informational.
